### ISSUE TITLE
License compliance for binary builds at Releases

### ISSUE THREAD

**Issue by ajorg-aws** [electron/electron#34236]
I'm trying to find corresponding source code and complete attributions for binary builds at [Releases](). For example https://github.com/electron/electron/releases/download/v18.2.3/electron-v18.2.3-darwin-x64.zip has an `ffmpeg.dylib` and FFmpeg has, at a minimum, an LGPL-2.1 license, which would require also distributing the "complete corresponding machine-readable source code" ([§4 of LGPL-2.1](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.en.html#:~:text=it%20with%20the-,complete%20corresponding%20machine%2Dreadable%20source%20code,-%2C%20which%20must%20be)) but I don't see this source code elsewhere in the Releases section (under Assets is where I'd expect to find it).

It looks like with sufficient sleuthing I might be able to piece it all together and find what is *likely* the complete corresponding source code, but it's much better when the organization that builds the software provides this so that we *know* that it's complete, and that it includes "the scripts used to control compilation and installation of the library." (§1, definition of "Source code"). That also ensures that the organization distributing the binaries is compliant with the license conditions.

Attributions (copyright statements and licenses) are also required by the licenses. I see most of those, for example I see `LICENSE` and `LICENSES.chromium.html` in the .zip packages, but I don't see anything for Squirrel and Mantle for example.

**Comment by ckerr** [comment]
disclaimer: IANAL. I'm a maintainer for Electron who has read the relevant licenses and this is my personal intepretatino of how the project complies.

## ffmpeg / LGPL-2.1

> FFmpeg has, at a minimum, an LGPL-2.1 license, which would require also distributing the "complete corresponding machine-readable source code" (§4 of LGPL-2.1) but I don't see this source code elsewhere in the Releases section (under Assets is where I'd expect to find it).

I think this is incorrect. Electron doens't make any modifications to FFmpeg; it simply is designed to work with the library by being compiled or linked with FFmpeg. So Electron's source is a "work that uses the library" as defined by LGPL-2.1 §5, and the Electron _executable_ is covered by §6:

> As an exception to the Sections above, you may also combine or link a "work that uses the Library" with the Library to produce a work containing portions of the Library, and distribute that work under terms of your choice, provided that the terms permit modification of the work for the customer's own use and reverse engineering for debugging such modifications.
>
> You must give prominent notice with each copy of the work that the Library is used in it and that the Library and its use are covered by this License. You must supply a copy of this License. If the work during execution displays copyright notices, you must include the copyright notice for the Library among them, as well as a reference directing the user to the copy of this License. Also, you must do one of these things:
>
> ... d) If distribution of the work is made by offering access to copy from a designated place, offer equivalent access to copy the above specified materials from the same place.

Electron's code is freely available, allowing modification / reverse engineering / etc. which satisfies the first requirement. Electron also satisfies clause 6d via `LICENSES.chromium.html` and also 6b because people can & do use system-preinstalled versions of ffmpeg with Electron. Either one of those would be sufficient to satisfy LGPL-2.1.

# Dependencies in Electron but not Chromium (Squirrel, Mantle, etc)

> Attributions (copyright statements and licenses) are also required by the licenses ... I don't see anything for Squirrel and Mantle for example.

This may be a fair point? I haven't looked closely so it's possible that there already is a notice somewhere that I'm missing. Regardless, it might be reasonable to be remove any doubt by adding a `LICENSES.electron.html` sibling to `LICENSES.chromium.html`. That file could include licenses for:

- Node.js
- nan
- squirrel.mac
- mantle
- ReactiveObjC

**Comment by ajorg-aws** [comment]
> I think this is incorrect. Electron doens't make any modifications to FFmpeg; it simply is designed to work with the library by being compiled or linked with FFmpeg. So Electron's source is a "work that uses the library"

Electron itself is a work that uses the library. I agree. And the distribution of Electron provided in Releases also distributes FFmpeg. It's that "also" that I was drawing your attention to. When you distribute a work under LGPL you need to also distribute the complete corresponding source code, including build scripts used to build it the way you built it. You can technically provide an "offer for source" but because Electron is something that your users will the redistribute, they would need to provide an offer for source that they cannot fulfill unless you fulfill the offer for source. So it's best for your users, and for you, if you provide that source code proactively instead.

Given this clarification, I've changed the issue title back. I hope you won't take that as rude.

**Comment by ckerr** [comment]
> It's that "also" that I was drawing your attention to. When you distribute a work under LGPL you need to also distribute the complete corresponding source code, including build scripts used to build it the way you built it.

No, I don't think it's rude! I'm glad that you're looking out for free software.

That said, I don't think that reading is correct. I quoted §6 in my previous reply because I think it's the relevant clause here: "as an exception to the Sections above" ... which overrides the requirement of section 4.

**Comment by ckerr** [comment]
Oh I think I'm talking past you. What I'm arguing about section 6 covers the released Electron binaries, e.g. `electron-v20.0.1-linux-x64.zip`, but you're talking about the fact that `ffmpeg-v20.0.1-linux-x64.zip` is also made available for download in the same release, and _that_ ffmpeg zipfile would be covered by §4.

Am I understanding you correctly?

**Comment by ajorg-aws** [comment]
Yes! That's the idea, except that `electron-v20.0.1-linux-x64.zip` *also* contains `libffmpeg.so`. I thought that `ffmpeg-v20.0.1-linux-x64.zip` contained a *different* build of FFmpeg (compiled with different options for licensing reasons?).

**Comment by ajorg-aws** [comment]
Oh, and it may not be only FFmpeg, because unless I'm mistaken there are some other copyleft things that are compiled into Chrome / Chromium / maybe Electron itself, and if they are statically compiled that may mean that while FFmpeg is a separate work, those copyleft licenses mean Electron needs to provide source code for all of Chromium as well.

I don't have details for all of that for the official builds, but I know there are lots of GPL and LGPL licenses mentioned in `LICENSES.chromium.html` that could be part of the build. It's unfortunate that Chromium doesn't adjust that file based on what's *actually* compiled in or not.

**Comment by MarshallOfSound** [comment]
The ffmpeg builds do not differ in licensing.  i.e. the license for ffmpeg in `LICENSES.chromium.html` is accurate for both builds.

The source code for Electron (and the build scripts required to reproduce the builds of both ffmpeg and Electron itself) are available at the tag in the GitHub Release, short of duplicate that code in _another_ zip file I'm not sure how else you'd like the code to be made available.  Electron and all of it's build scripts and dependencies are open-source.

My understanding of the licensing requirements of "providing source code" is just that the source code is accessible, which I believe we meet by being open source and tagging our build to source code.  All Chromium code is under the same obligations and is therefore also accessible as required by licensing.

To clarify the specific ask here is:
* Include licenses from Squirrel, Mantle, ReactiveObjC, NodeJS in the `LICENSES` files in the primary dist zip
* Is that all? The asks around source code don't make sense to me as all source code is already accessible and strongly tagged to a release

**Comment by ajorg-aws** [comment]
> The ffmpeg builds do not differ in licensing. i.e. the license for ffmpeg in `LICENSES.chromium.html` is accurate for both builds.

Not for copyright licensing, but for patent licensing, AIUI. This is the quote from the release notes that I think explains why there is a separate build:

> Provide prebuilt binaries of ffmpeg which do not include proprietary codecs.

It's also possible that copyright licensing is different between the two because switches in the configure flags determine if the build is GPL or GPL, version 2 or 3. I haven't looked at the configure flags used in the builds at Releases yet, so I'm not sure. Where do I find those?

> My understanding of the licensing requirements of "providing source code" is just that the source code is accessible, which I believe we meet by being open source and tagging our build to source code.

There's some nuance hiding in there. Licenses like MPL2 that's sufficient if unmodified, but GPL and LGPL ask the distributors themselves to provide the source code (even if unmodified). If we can accept that some users (and copyright holders) interpret the GPL this way, then we want to make it easier for them to comply.

> The asks around source code don't make sense to me as all source code is already accessible and strongly tagged to a release

I agree that it's possible to walk through the build scripts of a tag of Electron and spider out to all of the other source code required in order gather all of this together in order to comply with the licenses. The trouble is that while that takes care of the Electron project's obligations, at least in spirit if not following the letter of the (L)GPL, it leaves as an exercise for every developer who uses Electron to actually go through that process so that they can also comply with the licenses for their distributed products.

> short of duplicate that code in another zip file I'm not sure how else you'd like the code to be made available

I think that would help immensely. In fact I think it's ideal. And it's much easier for Electron to do that as part of the release process of Electron than it is for your users to piece it all together or rebuild the whole thing themselves.

Let me know if I can clarify further. Thank you so much for engaging on this topic and reaching the level of understanding we've already achieved. This gives me a lot of hope.

**Comment by MarshallOfSound** [comment]
> Not for copyright licensing, but for patent licensing, AIUI. This is the quote from the release notes that I think explains why there is a separate build:

Patent licensing of ffmpeg is a self-professed grey area. Ref: https://www.ffmpeg.org/legal.html. We provide the separate build out of an abundance of caution so that those who wish to avoid the grey area can.

>  I haven't looked at the configure flags used in the builds at Releases yet, so I'm not sure. Where do I find those?

The release GN args we use are here: https://github.com/electron/electron/blob/main/build/args/release.gn
The standalone ffmpeg is built using this: https://github.com/electron/electron/blob/main/build/args/ffmpeg.gn

If you want to find how those flags map to ffmpeg configure options you're going to have to do some _sleuthing_ through Chromium source code unfortunately.  It's been a long time since I've looked into this but my loose understanding from Back In The Day is that the license itself doesn't change by setting those flags.

> The trouble is that while that takes care of the Electron project's obligations, at least in spirit if not following the letter of the (L)GPL, it leaves as an exercise for every developer who uses Electron to actually go through that process so that they can also comply with the licenses for their distributed products.

That's my understanding too, as in, we are _technically_ compliant but could probably make it easier.  The workload for that though is not insignificant and given we're currently compliant there's no immediate need to go down that technical pit of multi-gigabyte zip files 😆 My immediate concern with issues like these is that any actual _issues_ are resolved, making things easier for people unfortunately takes a spot lower down the priority list.

I will say that the process of getting the code that Electron is built with isn't actually as labor intensive for a random person as you might think.  Checking out the right tag in `electron/electron` and running `gclient sync` will pull down every single dependency at the correct version in an identical way to what was used at build time 🤷

**Comment by ajorg-aws** [comment]
> I will say that the process of getting the code that Electron is built with isn't actually as labor intensive for a random person as you might think. Checking out the right tag in `electron/electron` and running `gclient sync` will pull down every single dependency at the correct version in an identical way to what was used at build time 🤷

I tried this after you mentioned it. It ran for 15 hours:
```txt
[15:15:49] Still working on:
[15:15:49]   src
```
I'm running it again now off-VPN and with fingers crossed (currently nearing an hour). The tool doesn't (by default?) indicate what if anything is going wrong.

I expect it will probably work eventually and then I should be able to figure out where exactly it put all that source code. I'm happy to write up and publish whatever short path I find to gathering the right source code, as an interim solution to this issue.

Because a release build necessarily must succeed at gathering source code, it should be easier to bundle and post it to Releases than for users of Electron, who are otherwise unfamiliar with the build process, to go through this themselves in order to ensure license compliance.

It also makes more sense for the project to do it once than for every developer who uses it to do it themselves, especially if the process is time and space consuming, and error prone in diverse environments as it seems to be for me right now.

**Comment by awjreynolds** [comment]
Apologies for jumping onto this but the implications are that when you build an Electron Application then the builder should include the Electron license but also the LICENSE.CHROMIUM.html attribution file and this should be part of any installation. I checked my MS Teams installation and this does not include the chromium license however my VS Code installation does (MacOS). Both installations have libffmpeg.dylib

**Comment by ajorg** [comment]
Unless something has been done to address this without the issue being referenced, I assume this issue is still relevant.

**Comment by awjreynolds** [comment]
I'd like to bring something into the mix that might be of concern. The FFMPEG library that "ships" with the electron installer that you build, contains proprietary codecs. Somebody has provided a work around https://github.com/MarshallOfSound/electron-packager-plugin-non-proprietary-codecs-ffmpeg HOWEVER this work around replaces FFMPEG post install with a "safe" FFMPEG build. Legally your application installer still ships with the proprietary codecs. I realise this is getting into a very legal area, but I'm unsure if the team behind Chromium somehow fixed the licensing so this is not a legal grey area for people shipping Electron solutions.

**Comment by ajorg** [comment]
I don't have any reason to believe this has been addressed. Un-stale for now.

**Comment by ajorg** [comment]
Not stale.

**Comment by p--b** [comment]
I don't think this is stale.

**Comment by ajorg** [comment]
Not stale

**Comment by ajorg** [comment]
Still not stale as far as anyone knows.

**Comment by ajorg** [comment]
Still unresolved, still matters.

**Comment by ajorg** [comment]
Unless something has changed, I expect this is still an issue. 

Would be nice (and compliant) to see some activity on it.

**Comment by ajorg** [comment]
😮‍💨

**Comment by p--b** [comment]
This is getting very dull. Please can we turn this bot off for this ticket? This will not be stale until it is addressed.
